### PR TITLE
refactor: #CRRE-465 refactoring and using model in structure group

### PR DESCRIPTION
**Pull request by BastienBenard**
## Describe your changes
Refactoring and using model in structure group.

API doc: https://confluence.support-ent.fr/display/CRRE/API+Structure+Group#APIStructureGroup-Deletegroupsofstructures

Added a dependency to pass the tests
org.powermock:powermock-core:2.0.2
org.powermock:powermock-api-mockito2:2.0.2
org.powermock:powermock-module-junit4:2.0.2
org.powermock:powermock-module-testng:2.0.2
io.vertx:vertx-codegen:3.9.5

## Checklist tests
Create, modify, delete a group. The logs in the log tab should show our changes.

## Issue ticket number and link
[CRRE-465](https://jira.support-ent.fr/browse/CRRE-465)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/crre/helpers/elasticsearch/ElasticSearch.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public void init(Vertx vertx, ConfigElasticSearch elasticsearchConfig) {
 		this.vertx = vertx;
 		List<String> serverUris = elasticsearchConfig.getServerUris();
 		String serverUri = elasticsearchConfig.getServerUri();
-		if (serverUris == null && serverUri != null) {
+		if ((serverUris == null || serverUris.isEmpty()) && serverUri != null) {
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/helpers/elasticsearch/ElasticSearch.java:88]: serverUris ne peut plus être null avec le ConfigModel mais il peut être vide

File: src/main/java/fr/openent/crre/model/config/ConfigElasticSearch.java [hunk 2]
```diff
@@ -20,7 +20,7 @@ public class ConfigElasticSearch implements IModel<ConfigElasticSearch> {
     private final boolean elasticSearchSsl;
 
     public ConfigElasticSearch(JsonObject jsonObject) {
-        this.serverUris = jsonObject.getJsonArray(ConfigField.SERVER_DASH_URIS).stream()
+        this.serverUris = jsonObject.getJsonArray(ConfigField.SERVER_DASH_URIS, new JsonArray()).stream()
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/model/config/ConfigElasticSearch.java:23]: Évite un NPE

File: src/main/resources/public/ts/model/StructureGroup.ts [hunk 3]
```diff
@@ -114,8 +114,5 @@ export class StructureGroups extends Selection<StructureGroup> {
     async sync() {
         let {data} = await http.get(`/crre/structure/groups`);
         this.all = Mix.castArrayAs(StructureGroup, data);
-        this.all.map((structureGroup) => {
-            structureGroup.structures = JSON.parse(structureGroup.structures.toString());
-        });
```
Comment (BastienBenard) [inline src/main/resources/public/ts/model/StructureGroup.ts:119]: L'API retourne directement un JsonArray maintenant

File: src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java [hunk 4]
```diff
@@ -3,85 +3,84 @@
...
+                .onFailure(promise::fail);
+
+        return promise.future();
     }
 
-    private void getStudentsByStructures(JsonArray structures) {
-        structureService.insertStudentsInfos(structures, event -> {
+    //Todo revoir les logs
+    private void getStudentsByStructures(List<String> structures) {
+        structureService.insertStudentsInfos(new JsonArray(new ArrayList<>(structures)), event -> {
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java:83]: Ce sera fait lors de la refacto de structureService

**Review by sjollois — APPROVED** [review]

File: src/main/java/fr/openent/crre/helpers/IModelHelper.java [hunk 5]
```diff
@@ -109,4 +117,21 @@ else if (validJsonClasses.stream().anyMatch(aClass -> aClass.isInstance(object))
...
     }
+
+    public static <T extends IModel<T>> Handler<Either<String, JsonArray>> sqlResultToIModel(Promise<List<T>> promise, Class<T> modelClass) {
+        return sqlResultToIModel(promise, modelClass, null);
+    }
+
+    public static <T extends IModel<T>> Handler<Either<String, JsonArray>> sqlResultToIModel(Promise<List<T>> promise, Class<T> modelClass, String errorMessage) {
+        return event -> {
+            if (event.isLeft()) {
+                if (errorMessage != null) {
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/helpers/IModelHelper.java:128]: Donc si tu n'as pas de message tu ne log rien ?
Ca me parait dangereux quand même.

File: src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java [hunk 6]
```diff
@@ -3,85 +3,84 @@
...
+        Future<Integer> getNextValFuture = this.getNextVal();
+        Future<List<String>> getOldIdStructureListFuture = this.getOldIdStructureList();
+
+        CompositeFuture.all(Arrays.asList(getNextValFuture, getOldIdStructureListFuture))
+                .compose(actualIdStructureList -> {
+                    Integer nextVal = getNextValFuture.result();
+                    List<TransactionElement> statements = new ArrayList<>();
+                    statements.add(getStructureGroupCreationStatement(nextVal, structureGroup));
+                    statements.add(getGroupStructureRelationshipStatement(nextVal, structureGroup.getStructures()));
+                    String errorMessage = String.format("An error occurred when launching transaction from %s::create", this.getClass().getSimpleName());
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java:65]: flag sur ton erreur

File: src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java [hunk 7]
```diff
@@ -90,45 +89,59 @@ private void getStudentsByStructures(JsonArray structures) {
...
+        Future<List<String>> oldIdStructureListFuture = this.getOldIdStructureList();
+
+        oldIdStructureListFuture
+                .compose(oldIdStructureList -> {
+
+                    List<TransactionElement> statements = new ArrayList<>();
+                    statements.add(getStructureGroupUpdateStatement(id, structureGroup));
+                    statements.add(getStructureGroupRelationshipDeletion(id));
+                    statements.add(getGroupStructureRelationshipStatement(id, structureGroup.getStructures()));
+                    String errorMessage = String.format("An error occurred when launching transaction from %s::update", this.getClass().getSimpleName());
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java:111]: flag sur ton erreur

File: src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java [hunk 8]
```diff
@@ -90,45 +89,59 @@ private void getStudentsByStructures(JsonArray structures) {
...
+            promise.complete(new JsonObject());
+            return promise.future();
+        }
 
-        sql.transaction(statements, event -> handler.handle(SqlQueryUtils.getTransactionHandler(event,ids.get(0))));
+        List<TransactionElement> transactionElementList = new ArrayList<>();
+        transactionElementList.add(getStructureGroupRelationshipDeletion(ids));
+        transactionElementList.add(getStructureGroupDeletion(ids));
+
+        String errorMessage = String.format("An error occurred when launching transaction from %s::delete", this.getClass().getSimpleName());
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/service/impl/DefaultStructureGroupService.java:139]: flag

**Review by qpic-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/crre/helpers/IModelHelper.java (hunk 5 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/helpers/IModelHelper.java:128]: C'est voulu, le développeur en question peut ne pas vouloir de log par cette fonction car il peut avoir une gestion des logs lorsque la promise fail a un autre endroit.

**Review by qpic-cgi — APPROVED** [review]

**Pull request merged**